pkg/logql/shardmapper.go

```diff
@@ -0,0 +1,224 @@
+			grouping:  expr.grouping,
+			params:    expr.params,
+			operation: expr.operation,
+		}, nil
+
+	case OpTypeAvg:
+		// avg(x) -> sum(x)/count(x)
+		lhs, err := m.mapVectorAggregationExpr(&vectorAggregationExpr{
+			left:      expr.left,
+			grouping:  expr.grouping,
+			operation: OpTypeSum,
+		})
+		if err != nil {
+			return nil, err
+		}
+		rhs, err := m.mapVectorAggregationExpr(&vectorAggregationExpr{
+			left:      expr.left,
+			grouping:  expr.grouping,
+			operation: OpTypeCount,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		return &binOpExpr{
+			SampleExpr: lhs,
+			RHS:        rhs,
+			op:         OpTypeDiv,
+		}, nil
+
+	case OpTypeCount:
+		// count(x) -> sum(count(x, shard=1) ++ count(x, shard=2)...)
+		sharded := m.mapSampleExpr(expr)
+		return &vectorAggregationExpr{
+			left:      sharded,
+			grouping:  expr.grouping,
+			operation: OpTypeSum,
+		}, nil
+	default:
+		// this should not be reachable. If an operation is shardable it should
+		// have an optimization listed.
+		level.Warn(util.Logger).Log(
+			"msg", "unexpected operation which appears shardable, ignoring",
+			"operation", expr.operation,
+		)
+		return expr, nil
+	}
+}
+
+func (m ShardMapper) mapRangeAggregationExpr(expr *rangeAggregationExpr) SampleExpr {
+	switch expr.operation {
+	case OpTypeCountOverTime, OpTypeRate:
+		// count_over_time(x) -> count_over_time(x, shard=1) ++ count_over_time(x, shard=2)...
+		// rate(x) -> rate(x, shard=1) ++ rate(x, shard=2)...
+		return m.mapSampleExpr(expr)
+	default:
+		return expr
+	}
+}
+
+// isShardable returns false if any of the listed operation types are not shardable and true otherwise
+func isShardable(ops []string) bool {
+	for _, op := range ops {
+		if shardable := shardableOps[op]; !shardable {
+			return false
+		}
+	}
+	return true
+}
+
+// shardableOps lists the operations which may be sharded.
+// topk, botk, max, & min all must be concatenated and then evaluated in order to avoid
+// potential data loss due to series distribution across shards.
+// For example, grouping by `cluster` for a `max` operation may yield
+// 2 results on the first shard and 10 results on the second. If we prematurely
+// calculated `max`s on each shard, the shard/label combination with `2` may be
+// discarded and some other combination with `11` may be reported falsely as the max.
+//
+// Explanation: this is my (owen-d) best understanding.
+//
+// For an operation to be shardable, first the sample-operation itself must be associative like (+, *) but not (%, /, ^).
+// Secondly, if the operation is part of a vector aggregation expression or utilizes logical/set binary ops,
+// the vector operation must be distributive over the sample-operation.
+// This ensures that the vector merging operation can be applied repeatedly to data in different shards.
+// references:
+// https://en.wikipedia.org/wiki/Associative_property
+// https://en.wikipedia.org/wiki/Distributive_property
+var shardableOps = map[string]bool{
```

Comment:
nit: this could be a map[string]struct{}{

Reply:
sure, but then I'd have to do `_, ok :=` instead of just `shardable = `.